File: python/src/agntz/client/events.py
"""Normalize worker SSE frames into Python client events."""

from __future__ import annotations

import json
from typing import Any

from ._sse import SseFrame
from .errors import StreamError
from .models import Event


def normalize_agent_event(frame: SseFrame) -> Event | None:
    if frame.event is None:
        return None
    payload = _parse_payload(frame)
    if frame.event == "run-start":
        return Event(
            type="start",
            agentId=_string(payload, "agentId"),
            kind=_agent_kind(payload),
            sessionId=_string(payload, "sessionId"),
        )
    if frame.event == "run-complete":
        state = payload.get("state")
        return Event(
            type="complete",
            output=payload.get("output"),
            state=state if isinstance(state, dict) else {},
            sessionId=_string(payload, "sessionId"),
        )
    if frame.event == "run-error":
        return Event(type="error", error=_string(payload, "error"))
    if frame.event == "reply":
        seq = payload.get("seq")
        return Event(
            type="reply",
            text=_string(payload, "text"),
            ts=_string(payload, "ts"),
            sessionId=_string(payload, "sessionId"),
            runId=_string(payload, "runId"),
            seq=seq if isinstance(seq, int) else None,
        )
    return None


def normalize_run_event(frame: SseFrame) -> Event | None:
    if frame.event is None:
        return None
    payload = _parse_payload(frame)
    if frame.event in {
        "run-spawn",
        "text-delta",
        "tool-call-start",
        "tool-call-end",
        "step-complete",
        "draining",
        "reply",
        "run-complete",
        "run-error",
        "run-cancelled",
    }:
        return Event.model_validate(payload)
    if frame.event == "snapshot":
        return Event.model_validate({"type": "snapshot", "run": payload})
    if frame.event == "stream-error":
        raise StreamError(_string(payload, "error"), code="STREAM_ERROR")
    return None


def normalize_trace_event(frame: SseFrame) -> Event | None:
    if frame.event is None:
        return None
    payload = _parse_payload(frame)
    if frame.event == "span-start":
        return Event.model_validate({"type": "span-start", "span": payload.get("span")})
    if frame.event == "span-end":
        return Event.model_validate(
            {
                "type": "span-end",
                "spanId": _string(payload, "spanId"),
                "patch": payload.get("patch") or {},
            }
        )
    if frame.event == "trace-done":
        return Event.model_validate({"type": "trace-done", "summary": payload.get("summary")})
    if frame.event == "snapshot":
        return Event.model_validate(
            {
                "type": "snapshot",
                "summary": payload.get("summary"),
                "spans": payload.get("spans") or [],
            }
        )
    if frame.event == "stream-error":
        raise StreamError(_string(payload, "error"), code="STREAM_ERROR")
    return None


def _parse_payload(frame: SseFrame) -> dict[str, Any]:
    try:
        value = json.loads(frame.data)
    except json.JSONDecodeError as exc:
        raise StreamError(
            f'Invalid JSON in "{frame.event}" event data',
            code="INVALID_SSE_PAYLOAD",
            cause=exc,
        ) from exc
    if not isinstance(value, dict):
        raise StreamError("SSE event data must be a JSON object", code="INVALID_SSE_PAYLOAD")
    return value


def _string(payload: dict[str, Any], field: str) -> str:
    value = payload.get(field)
    if not isinstance(value, str):
        raise StreamError(
            f'SSE payload missing string field "{field}"',
            code="INVALID_SSE_PAYLOAD",
        )
    return value


def _agent_kind(payload: dict[str, Any]) -> str:
    value = payload.get("kind")
    if value in {"llm", "tool", "sequential", "parallel"}:
        return value
    raise StreamError(f"Unknown agent kind: {value}", code="INVALID_SSE_PAYLOAD")

File: packages/client/src/events.ts
import { StreamError } from "./errors.js";
import type {
	AgentKind,
	MultiplexedRunEvent,
	Run,
	Span,
	SseFrame,
	StreamEvent,
	TraceLiveEvent,
	TraceSummary,
} from "./types.js";

/**
 * Map an SSE wire frame to a public StreamEvent. Unknown events return null
 * so the wire format can evolve without breaking consumers. Invalid JSON in
 * the data payload throws StreamError — that's a real protocol violation.
 */
export function normalizeEvent(frame: SseFrame): StreamEvent | null {
	if (!frame.event) return null;
	const payload = parseData(frame.data, frame.event);
	switch (frame.event) {
		case "run-start": {
			const agentId = asString(payload, "agentId");
			const kind = asAgentKind(payload);
			const sessionId = asString(payload, "sessionId");
			return { type: "start", agentId, kind, sessionId };
		}
		case "run-complete": {
			const output = (payload as { output?: unknown }).output;
			const state = asStateRecord(payload);
			const sessionId = asString(payload, "sessionId");
			return { type: "complete", output, state, sessionId };
		}
		case "run-error": {
			const error = asString(payload, "error");
			return { type: "error", error };
		}
		case "reply": {
			const text = asString(payload, "text");
			const ts = asString(payload, "ts");
			const sessionId = asString(payload, "sessionId");
			const runId = asString(payload, "runId");
			const seqVal = (payload as { seq?: unknown }).seq;
			const seq = typeof seqVal === "number" ? seqVal : undefined;
			return seq === undefined
				? { type: "reply", text, ts, sessionId, runId }
				: { type: "reply", text, ts, sessionId, runId, seq };
		}
		default:
			return null;
	}
}

function parseData(data: string, event: string): unknown {
	try {
		return JSON.parse(data);
	} catch (cause) {
		throw new StreamError(`Invalid JSON in "${event}" event data`, {
			code: "INVALID_SSE_PAYLOAD",
			cause,
		});
	}
}

function asString(payload: unknown, field: string): string {
	const obj = payload as Record<string, unknown> | null;
	const value = obj?.[field];
	if (typeof value !== "string") {
		throw new StreamError(`SSE payload missing string field "${field}"`, {
			code: "INVALID_SSE_PAYLOAD",
		});
	}
	return value;
}

function asAgentKind(payload: unknown): AgentKind {
	const kind = (payload as { kind?: unknown }).kind;
	if (
		kind === "llm" ||
		kind === "tool" ||
		kind === "sequential" ||
		kind === "parallel"
	) {
		return kind;
	}
	throw new StreamError(`Unknown agent kind: ${String(kind)}`, {
		code: "INVALID_SSE_PAYLOAD",
	});
}

function asStateRecord(payload: unknown): Record<string, unknown> {
	const state = (payload as { state?: unknown }).state;
	if (state && typeof state === "object" && !Array.isArray(state)) {
		return state as Record<string, unknown>;
	}
	return {};
}

/**
 * Map an SSE wire frame from /runs/:id/stream to a typed MultiplexedRunEvent.
 * Returns null for unknown event types so the wire format can evolve. Throws
 * StreamError on invalid JSON or shape violations.
 */
export function normalizeRunEvent(frame: SseFrame): MultiplexedRunEvent | null {
	if (!frame.event) return null;
	const payload = parseData(frame.data, frame.event);
	switch (frame.event) {
		case "run-spawn":
		case "text-delta":
		case "tool-call-start":
		case "tool-call-end":
		case "step-complete":
		case "draining":
		case "reply":
		case "run-complete":
		case "run-error":
		case "run-cancelled":
			return payload as MultiplexedRunEvent;
		case "snapshot": {
			return { type: "snapshot", run: payload as Run };
		}
		case "stream-error": {
			const error = asString(payload, "error");
			throw new StreamError(error, { code: "STREAM_ERROR" });
		}
		default:
			return null;
	}
}

/**
 * Map an SSE wire frame from /traces/:id/stream to a typed TraceLiveEvent.
 * The `snapshot` event is SDK-only — the registry never emits it; the worker
 * synthesises it for terminal traces in the stream endpoint.
 */
export function normalizeTraceLiveEvent(
	frame: SseFrame,
): TraceLiveEvent | null {
	if (!frame.event) return null;
	const payload = parseData(frame.data, frame.event);
	switch (frame.event) {
		case "span-start": {
			const span = (payload as { span?: unknown }).span as Span;
			return { type: "span-start", span };
		}
		case "span-end": {
			const obj = payload as { spanId?: unknown; patch?: unknown };
			const spanId = asString(obj, "spanId");
			const patch = (obj.patch as Partial<Span>) ?? {};
			return { type: "span-end", spanId, patch };
		}
		case "trace-done": {
			const summary = (payload as { summary?: unknown })
				.summary as TraceSummary;
			return { type: "trace-done", summary };
		}
		case "snapshot": {
			const obj = payload as { summary?: unknown; spans?: unknown };
			const summary = obj.summary as TraceSummary;
			const spans = (obj.spans as Span[]) ?? [];
			return { type: "snapshot", summary, spans };
		}
		case "stream-error": {
			const error = asString(payload, "error");
			throw new StreamError(error, { code: "STREAM_ERROR" });
		}
		default:
			return null;
	}
}
